flake.nix

```diff
@@ -0,0 +1,156 @@
+            , llvm-pretty-src
+            , fgl-src
+            , fgl-visualize-src
+            }:
+    let
+      shellWith = pkgs: adds: drv: drv.overrideAttrs(old:
+        { buildInputs = old.buildInputs ++ adds pkgs; });
+      # Add additional packages useful for a development shell, generally
+      # representing test packages or non-propagated build dependencies of
+      # various sub-packages.
+      shellPkgs = pkgs: [
+        # pkgs.haskell.compiler.integer-simple.ghc8107
+        # pkgs.haskell.packages.ghc8107.profiteur
+        pkgs.cabal-install
+        pkgs.llvm_11
+        pkgs.clang_11
+      ];
+    in rec {
+      devShells =
+        let oneshell = s: n:
+              let pkgs = import nixpkgs { system=s; };
+              in levers.variedTargets
+                { ghcver = levers.validGHCVersions pkgs.haskell.compiler; }
+                ( { ghcver, ... } @ vargs:
+                  shellWith pkgs shellPkgs
+                    (self.packages.${s}.${n}.${ghcver}.env.overrideAttrs (a:
+                      {
+                        # Set envvars here
+                      }
+                    )));
+        in levers.eachSystem
+          (s:
+            let pkgs = import nixpkgs { system=s; };
+                names = builtins.attrNames (self.packages.${s});
+                outs = builtins.removeAttrs
+                  (pkgs.lib.genAttrs names (oneshell s))
+                  [ "ghc" ];
+                shells = pkgs.lib.attrsets.mapAttrs (n: v: v.default) outs;
+            in shells
+          ) ;
+
+      packages = levers.eachSystem (system:
+        let
+          mkHaskell = levers.mkHaskellPkg {
+            inherit nixpkgs system;
+            };
+          pkgs = import nixpkgs { inherit system; };
+          wrap = levers.pkg_wrapper system pkgs;
+          haskellAdj = drv:
+            with (pkgs.haskell).lib;
+            dontHaddock (dontCheck (dontBenchmark (drv)));
+          llvm-pretty-bc-parser-test = built: llvm:
+            derivation {
+              inherit system;
+              name = "llvm-pretty-bc-parser-test-with-llvm${llvm.version}";
+              builder = "${pkgs.bash}/bin/bash";
+              args = [ "-c"
+                       ''
+                       ${pkgs.coreutils}/bin/cp -r ${built}/test-build/* .
+                       export PATH=${llvm}/bin:${pkgs.diffutils}/bin:$PATH
+                       ./dist/build/unit-test/unit-test
+                       ./dist/build/disasm-test/disasm-test
+                       echo OK > $out
+                       ''
+                     ];
+              buildInputs = [ llvm pkgs.diffutils pkgs.coreutils ];
+            };
+        in rec {
+          default = llvm-pretty-bc-parser;
+          TESTS = wrap "llvm-pretty-bc-parser-TESTS"
+            (builtins.map
+              (llvm-pretty-bc-parser-test llvm-pretty-bc-parser-test-build)
+              [
+                pkgs.llvm_10
+                pkgs.llvm_11
```

Comment:
This is where all of the different LLVM versions are enumerated, right? It might be worth stating this somewhere in a comment—I had to trawl through the CI logs before I realized which LLVM versions were actually being tested.

Relatedly, is there any reason we couldn't go ahead and add more versions here?

Reply:
Hrm.. I thought I had replied to this, but I don't see it now: apologies if this ends up being a duplicate.

Short version is: there are issues with both lower and higher LLVM versions that will probably want accompanying fixes (in both code and modifications to testing).  These seem to be the two "stable"/"baseline" versions that are good to start with, and this PR was already getting large, so I decided to leave this range expansion for future PRs.

Reply:
And yes, this is the location where the LLVM versions are specified and I added a commit with descriptive comments.